server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java

```diff
@@ -0,0 +1,179 @@
+
+        ShardSnapshotTask(SnapshotShardContext context) {
+            this.context = context;
+        }
+
+        public SnapshotShardContext context() {
+            return context;
+        }
+
+        public void run() {
+            shardSnapshotter.accept(context);
+        }
+
+        public short priority() {
+            return 1;
+        }
+
+        @Override
+        public int compareTo(ShardSnapshotTask other) {
+            int res = context.snapshotId().compareTo(other.context.snapshotId());
+            if (res != 0) {
+                return res;
+            }
+            return Integer.compare(priority(), other.priority());
+        }
+    }
+
+    class FileSnapshotTask extends ShardSnapshotTask {
+        private final FileInfo fileInfo;
+        private final ActionListener<Void> fileUploadListener;
+
+        FileSnapshotTask(SnapshotShardContext context, FileInfo fileInfo, ActionListener<Void> fileUploadListener) {
+            super(context);
+            this.fileInfo = fileInfo;
+            this.fileUploadListener = fileUploadListener;
+        }
+
+        @Override
+        public void run() {
+            ActionRunnable.run(fileUploadListener, () -> fileSnapshotter.accept(context, fileInfo)).run();
+        }
+
+        @Override
+        public short priority() {
+            return 2;
+        }
+    }
+
+    public ShardSnapshotWorkers(
+        final int maxWorkers,
+        final Executor executor,
+        final Consumer<SnapshotShardContext> shardSnapshotter,
+        final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter
+    ) {
+        assert maxWorkers > 0;
+
+        logger.info("starting shard snapshot worker pool of max size {}", maxWorkers);
+        this.maxWorkers = maxWorkers;
+        this.executor = executor;
+        this.shardSnapshotter = shardSnapshotter;
+        this.fileSnapshotter = fileSnapshotter;
+    }
+
+    public void enqueueShardSnapshot(final SnapshotShardContext context) {
+        logger.trace("enqueuing shard snapshot task [snapshotID={}, indexID={}]", context.snapshotId(), context.indexId());
+        synchronized (mutex) {
+            snapshotTasks.add(new ShardSnapshotTask(context));
+            ensureEnoughWorkers();
+        }
+    }
+
+    public void enqueueFileUpload(
+        final SnapshotShardContext context,
+        final FileInfo fileInfo,
+        final ActionListener<Void> fileUploadListener
+    ) {
+        logger.trace(
+            "enqueuing shard snapshot file upload task [snapshotID={}, indexID={}, file={}]",
+            context.snapshotId(),
+            context.indexId(),
+            fileInfo.name()
+        );
+        synchronized (mutex) {
+            snapshotTasks.add(new FileSnapshotTask(context, fileInfo, fileUploadListener));
+            ensureEnoughWorkers();
+        }
+    }
+
+    private void ensureEnoughWorkers() {
+        assert Thread.holdsLock(mutex);
```

Comment:
I wonder if this method would be simpler to reason about if it was lock-free, i.e., using an `AtomicInteger` for the worker count and just spawn one new thread if there is budget.